tools/bootnode/bootnode.go

```diff
@@ -46,56 +46,53 @@ func main() {
 		logging.SetDebugLogging()
 	}
 
-	listen, err := ma.NewMultiaddr(fmt.Sprintf("/ip4/0.0.0.0/tcp/%d", *port))
-	if err != nil {
-		log.Fatalf("Failed to construct new multiaddress. %v", err)
-	}
+	defaultIP := "0.0.0.0"
 
-	opts := []libp2p.Option{
-		libp2p.ListenAddrs(listen),
-	}
-	opts = addPrivateKeyOpt(opts)
+	privKey := extractPrivateKey()
+	listener := createListener(defaultIP, *port, privKey)
 
-	ctx := context.Background()
+	node := listener.Self()
+	fmt.Printf("Running bootnode: /ip4/%s/udp/%d/discv5/%s\n", node.IP.String(), node.UDP, node.ID.String())
 
-	host, err := libp2p.New(ctx, opts...)
-	if err != nil {
-		log.Fatalf("Failed to create new host. %v", err)
-	}
+	select {}
+}
 
-	dopts := []dhtopts.Option{
-		dhtopts.Datastore(dsync.MutexWrap(ds.NewMapDatastore())),
-		dhtopts.Protocols(
-			protocol.ID(dhtProtocol),
-		),
+func createListener(ipAddr string, port int, privKey *ecdsa.PrivateKey) *discv5.Network {
+	udpAddr := &net.UDPAddr{
+		IP:   net.ParseIP(ipAddr),
+		Port: port,
 	}
-
-	dht, err := kaddht.New(ctx, host, dopts...)
+	conn, err := net.ListenUDP("udp4", udpAddr)
 	if err != nil {
-		log.Fatalf("Failed to create new dht: %v", err)
+		log.Fatal(err)
 	}
-	if err := dht.Bootstrap(context.Background()); err != nil {
-		log.Fatalf("Failed to bootstrap DHT. %v", err)
-	}
-
-	fmt.Printf("Running bootnode: /ip4/0.0.0.0/tcp/%d/p2p/%s\n", *port, host.ID().Pretty())
 
-	select {}
+	network, err := discv5.ListenUDP(privKey, conn, "", nil)
+	if err != nil {
+		log.Fatal(err)
+	}
+	return network
 }
 
-func addPrivateKeyOpt(opts []libp2p.Option) []libp2p.Option {
+func extractPrivateKey() *ecdsa.PrivateKey {
+	var privKey *ecdsa.PrivateKey
+	var err error
 	if *privateKey != "" {
```

Comment:
how are we going to get into this code path?

Reply:
@shayzluf this is initalized by a flag, so if a flag is given we can use the given private key 
or just generate one in the other case

Reply:
sg